### PR TITLE
Add state module

### PR DESCRIPTION
**Pull request by xevisalle**

### CODE REVIEW

**Review by ureeves — APPROVED** [review]
LGTM, but one question - since we're implementing citadel here in some abstract way, could it be valuable to use an arbitrary tree depth?

File: src/state.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        let lpk = JubJubAffine::from(*lic.lsa.pk_r().as_ref());
+        let license_hash = sponge::hash(&[lpk.get_x(), lpk.get_y()]);
+
+        self.tree.push(DataLeaf::new(license_hash));
+    }
+    pub fn get_merkle_proof(&self, license_hash: &BlsScalar) -> PoseidonBranch<DEPTH> {
+        let mut pos = 0;
+
+        for i in 0..(4 ^ DEPTH) {
+            let it = i.try_into().unwrap();
```
Comment (miloszm) [inline src/state.rs:40]: `try_into()` followed by `unwrap()` boils down to the same as `into()`, it panics when something goes wrong, so I'd rather use `into()`

File: src/state.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+            let leaf = self.tree.get(it);
+
+            match leaf {
+                Some(leaf) if leaf.license_hash == *license_hash => pos = it,
+                _ => (),
+            }
+        }
```
Comment (miloszm) [inline src/state.rs:48]: this loop has a huge range, `4 ^ DEPTH`, and nothing seems to interrupt it or finish it early, are we sure that is what we want? We could interrupt the loop when pos is found, why is this not done?

**Review by miloszm — CHANGES_REQUESTED** [review]

File: src/state.rs (hunk 1 above)
Comment (xevisalle) [reply src/state.rs:40]: good one, I'll fix it!

File: src/state.rs (hunk 2 above)
Comment (xevisalle) [reply src/state.rs:48]: good question, I removed the break for testing and I forgot to readd it

**Review by miloszm — APPROVED** [review]

**Pull request merged**